Ionic 4 support

captaincaius opened the pull request.

captaincaius commented:
So far, everything I tested is functional and generated tests pass, but in looking at ionic 4's activity, we probably need to wait until ionic stabilizes for the final PR.  There's already big changes between master and the latest beta as far as styling goes as well as the new NavController's interface (the facade for angular's router).  That's why it's currently pretty ugly ;) - but it looks like they're settling on how to properly set up the theme defaults.  Hopefully by the next beta this will be pretty again.

Since it's a huge PR, at least this will give you a chance to review the direction of the majority of the changes involved.

I'll try to annotate some of the specifics tomorrow.

captaincaius commented on generators/app/templates/.editorconfig:
```diff
@@ -8,7 +8,7 @@ indent_size = 2
 insert_final_newline = true
 trim_trailing_whitespace = true
 end_of_line = lf
```
This is what the ionic starters have, and I didn't want anything I borrowed from ionic to flag here

captaincaius commented:
quick update - the protractor issue above (and the wait) was solved in the last commit.  the e2e tests were refreshing the page like 5 times.

sinedied commented:
Thanks for this huge contribution! I'll take a look at the changes as soon as I can and will reply to your question, but my wife gave birth yesterday so it will be delayed a little 😄 

Meanwhile, @luiza-cicone and @bursauxa can also review your changes 😉

bursauxa commented:
I will have to give this a pass as I have zero experience with ionic.

The one thing I can say is that we should not change our style guidelines because of another library (referring specifically to the 120 characters per line that @captaincaius raised to 140). I mean, maybe 140 is a better limit than 120, but then it should be our decision, not collateral damage from code that was obtained from ionic.

captaincaius commented:
Wow, congrats on the newborn @sinedied !!  :fireworks: 

@bursauxa thanks for the feedback, and thanks for updating the branch! :)

About the character limit, yeah I totally agree.  I'll revert it and deal with any side-effects (heck, there may not be any), and then take it off the list above :).  I'll wait on feedback from one more review and do it along with any other fixes needed.

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.html:
```diff
@@ -1,12 +1,14 @@
-          <button ion-button icon-only (click)="changeLanguage()">
+          <ion-button class="profile-button" color="light" (click)="showProfileActions()"
+            round fill="outline" size="small" slot="start"
+          >
```
`>` should be on the previous line to keep it consistent with existing code style

captaincaius replied:
moved :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.ts:
```diff
@@ -1,7 +1,8 @@
+import { Router } from '@angular/router';
+import { ActionSheetController, AlertController, Platform } from '@ionic/angular';
+import { ActionSheetButton, ActionSheetOptions } from '@ionic/core';
+import { TextFieldTypes } from '@ionic/core';
```
factorize import

captaincaius replied:
fixed :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.ts:
```diff
@@ -32,19 +32,9 @@ export class ShellComponent implements OnInit {
-  showProfileActions() {
-    const actionSheetOptions: ActionSheetOptions = { title: this.username || undefined };
-    const actionSheet = this.actionSheetController.create(actionSheetOptions);
+  async showProfileActions() {
```
same remark for async/await

captaincaius replied:
done :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.ts:
```diff
@@ -92,19 +87,19 @@ export class ShellComponent implements OnInit {
   }
 
-  private changeLanguage() {
+  private async changeLanguage() {
```
same for async/await

captaincaius replied:
done :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.ts:
```diff
@@ -92,19 +87,19 @@ export class ShellComponent implements OnInit {
   }
 
-  changeLanguage() {
+  async changeLanguage() {
```
same for async/await

captaincaius replied:
done :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-side-menu._shell.component.ts:
```diff
@@ -92,19 +87,19 @@ export class ShellComponent implements OnInit {
+        header: this.translateService.instant('Change language'),
         inputs: this.i18nService.supportedLanguages.map(language => ({
-          type: 'radio',
+          type: 'radio' as TextFieldTypes,
```
`TextFieldTypes.Radio`?

captaincaius replied:
replied on other comment :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-tabs.shell.component.html:
```diff
@@ -1,3 +1,13 @@
+    show="true"
+    [active]="tabIndex==selectedTabIndex"
+    [selected]="tabIndex==selectedTabIndex"
+  >
```
formatting not consistent with the rest

captaincaius replied:
joined :)

sinedied commented on generators/app/templates/src/app/shell/__ionic-tabs.shell.component.ts:
```diff
@@ -13,31 +13,27 @@ import { HomeComponent } from '@app/home/home.component';
-  onTabChange(selectedTab: Tab) {
-    this.router.navigate([this.tabs[selectedTab.index].route]);
+  onTabChange(selectedTabElm: Tab) {
+    const selectedTab = find(this.tabs, { name: selectedTabElm.name });
```
use ES6 array method for `find` instead of lodash, as lodash will be removed from default libs in a coming commit

captaincaius replied:
ok will do :) I saw lodash being used below, so I tried to follow.

captaincaius replied:
de-lodashed :)

sinedied commented on generators/app/templates/src/theme/__ionic.theme-variables.scss:
```diff
@@ -32,54 +26,110 @@ $app-direction: ltr;
-  light:      #f4f4f4,
-  dark:       #333
-);
+:root {
```
is `:root` really necessary here?

captaincaius replied:
tbh I'm not sure, but their theming guides say to use :root, and their starters have it too.

sinedied commented on generators/app/templates/src/theme/__ionic.theme-variables.scss:
```diff
@@ -32,54 +26,110 @@ $app-direction: ltr;
 // --------------------------------------------------
 
-@import "~ionic-angular/fonts/roboto";
-@import "~ionic-angular/fonts/noto-sans";
```
comments related to the imports should be removed as well, they were from Ionic code 😉

captaincaius replied:
removed for now :) I expect them to come back when they expand upon theming.

sinedied commented on generators/app/templates/tslint.json:
```diff
@@ -30,7 +30,7 @@
     "max-line-length": [
       true,
-      120
+      140
```
should stick to default

captaincaius replied:
reverted :)

sinedied commented on scripts/test.sh:
```diff
@@ -58,17 +58,17 @@ do
-        npm run test:ci -- --no-progress
-        npm run e2e -- --no-progress
-        npm run build -- --no-progress
+        npm run test:ci --
```
`--progress` should only be (temporarily) disabled for e2e tests, after update to latest master the fix is already here

captaincaius replied:
awesome. ok.

captaincaius replied:
reverted :)

sinedied commented on scripts/tests/app/web/ionic-authentication.json:
```diff
@@ -4,6 +4,5 @@
   "layout": "side-menu",
   "auth": true,
-  "lazy": false,
-  "location": "hash"
```
keep hash option tested as it's needed in some cases

captaincaius replied:
reverted :) hash works on ionic now too

sinedied commented on scripts/tests/app/web/material.json:
```diff
@@ -5,6 +5,5 @@
   "auth": false,
   "lazy": false,
-  "angulartics": false,
-  "location": "hash"
```
same

captaincaius replied:
reverted :) I totally misread some of the generator code

captaincaius commented:
Thanks for all the details in the review!  I have a lot more clarity now on your goals.

And speaking of which, thanks for the vision update :)   I see you released 5.1.0, and it includes some angular CLI/etc updates.  I guess I should be able to undo some of the workarounds above (namely, the CLI flags vs angular.json stuff).  Looking through issues, it wasn't clear to me whether those flags were coming back or not.

I'll address each task individually, but to address your high-level comments:

- line lengths: gotcha :) already reverted locally after @bursauxa 's feedback - just didn't want to abuse your CI for that alone ;).
- HashLocationStrategy: Okay.  I'll wrestle with it and do some testing to make sure it also works before the next push.
- async/await stuff: good point - I'll change all the new promise-related stuff to use pure promises.  Sorry I didn't think about consistency - they changed some popover/modal/etc stuff to use promises, and out of habit I reached for async/await without looking around first ;).  I'll try to make the purely-promise version as readable as I can.
- Ionic-CLI-related things - Okay.  I thought that since they've finally turned their CLI stuff (lol shenanigans) into a thin layer over angular CLI, it would be easy to support it alongside, but I wasn't sure if you'd want to.  But okay, I'll rip it all out :).

It looks like the merge conflicts from 5.1.0 are pretty small, so I'll update soon and continue.  But I do think we should wait for another release or so from ionic so a few things are ironed out (visual quirks, etc).

captaincaius commented on generators/app/templates/_angular.json:
```diff
@@ -17,6 +17,8 @@
         "build": {
           "builder": "@angular-devkit/build-angular:browser",
           "options": {
+            "progress": false,
```
cool beans.  As you probably know, the most recent version of the CLI at the time didn't have those flags working, so I had to move it here.  But I see master is using an updated version, so I'm guessing I can revert it and use the flags instead :).

captaincaius replied:
reverted :)

captaincaius commented on generators/app/templates/_angular.json:
```diff
@@ -17,6 +17,8 @@
           "builder": "@angular-devkit/build-angular:browser",
           "options": {
+            "progress": false,
+            "aot": true,
```
ditto above.  I guess you fixed this via the dependency updates in master.  will remove.

captaincaius replied:
reverted :)

captaincaius commented on generators/app/templates/e2e/src/_app.po.ts:
```diff
@@ -7,9 +7,15 @@ import { browser, element, by } from 'protractor';
+  usernameField = element(by.css('ion-input[formControlName="username"]'));
+  passwordField = element(by.css('ion-input[formControlName="password"]'));
+  loginButton = element(by.css('ion-button[type="submit"]'));
+<%   } else { -%>
```
ah. good point :) thanks.

captaincaius replied:
I hoped that would work, but it seemed it was too broad - tests started failing, and when they were passing I'd sometimes see that "selector matched multiple elements" warning from protractor.

I'm guessing ionic slaps the attribute on the native input inside too or something.  I can investigate further if you like.

This also brings up a philosophical question :P...  I've never seen full consensus on best practice for this - do you think it's worthwhile to add test-specific attributes to the code, aside from what's already there?  These are the various things I've seen people do:

- don't add anything to the code just for tests - use what's already there (like we're doing here)
- only select based on test-designated attributes
- only select based on test-designated css classes (yeah I know - that's not what they're for)

captaincaius replied:
damn - that conversation belongs on that other issue - you can disregard this comment ;) I'll move it there

captaincaius replied:
ditto above - pending other PR conversation

sinedied replied:
Too bad it does not work, in that case disregard my comment 😞 

As for the philosophical question, I agree there's no clear consensus on this. My opinion is that we should avoid pushing one if we can, ie keep the motto `don't add anything to the code just for tests - use what's already there (like we're doing here)`.

If people wants to add test-specific attributes/classes, they can always add it layer (I saw some team adding `data-test="<test-id>"` attributes for e2e testing for ex).

captaincaius commented on generators/app/templates/e2e/src/_app.e2e-spec.ts:
```diff
@@ -6,20 +6,18 @@ import { AppPage } from './app.po';
   let page: AppPage;
 
-  beforeEach(() => {
+  beforeAll(() => {
```
Well these test cases aren't currently independent - it's really one test case, but split in half.  I'd looked and noticed you had a separate issue for overhauling the e2e tests to make them more extensive and be a good example on best practices, etc, so I figured changing the tests too much was outside the scope of this issue.

Because of that, I just did something quick and dirty to eliminate all the extra refreshes, as they were making the ionic version of login fail.  I didn't fully get to the bottom of it, but it seemed to be some kind of race condition.

I don't know how deep you want to go within this issue - another simple approach would be to revert this change, but combine the two it blocks.

captaincaius replied:
for this one I'm waiting for a thumbs-up or thumbs-down on the e2e PR I filed.  If you like it and merge it, it'll render these changes obsolete.

if not, let me know and I'll address these here :)

sinedied replied:
I approved your e2e PR, you can update this one after it's merged 😃

captaincaius commented on generators/app/templates/src/app/login/__auth._login.component.spec.ts:
```diff
@@ -3,7 +3,9 @@ import { TranslateModule } from '@ngx-translate/core';
 import { ReactiveFormsModule } from '@angular/forms';
 <% if (props.ui === 'ionic') { -%>
-import { IonicModule } from 'ionic-angular';
+import { FormsModule, FormBuilder } from '@angular/forms';
```
about FormsModule - yup, the language picker ;).

about FormBuilder - it seems you're right - sorry.

captaincaius commented on generators/app/templates/src/app/login/__auth._login.component.ts:
```diff
@@ -38,28 +38,37 @@ export class LoginComponent implements OnInit {
         log.debug(`Login error: ${error}`);
         this.error = error;
+<% if (props.ui === 'ionic') { -%>
+        loadingPresentedPromise.then(() => loading.dismiss());
```
I remember having considered exactly that, but decided I shouldn't for some reason.  I'll try to remember why when I have the full code in front of me and comment back (or just do it that way if it turns out I was wrong :P)

captaincaius replied:
fixed :) good catch

sinedied commented:
> But I do think we should wait for another release or so from ionic so a few things are ironed out (visual quirks, etc).

Good point! I was thinking of creating a new v6 branch that will contains the major changes for the next release (Ionic 4 and Electron for now) and release it as a beta under the npm tag `@next` like Angular preversions, what do you think?
This way we can iterate on the changes as Ionic 4 is still beta, while have a ready to install release for those that want to take the plunge 😄

luiza-cicone commented on generators/app/templates/src/app/login/__auth._login.component.ts:
```diff
@@ -40,16 +40,17 @@ export class LoginComponent implements OnInit {
         this.loginForm.markAsPristine();
 <% if (props.ui === 'ionic') { -%>
-        loading.dismiss();
+        loadingPresentedPromise.then(() => loadingPromise.then(loading => loading.dismiss()));
```
I think you could chain all the async work. You can transform a promise into an observable to keep the flow and also avoid saving intermediate promises.

could this work?:
```
const login$ = this.authenticationService.login(this.loginForm.value));
<% if (props.ui === 'ionic') { -%>
from(this.loadingController.create());
<% } else { -%>
login$
<% } -%>
    .pipe(
<% if (props.ui === 'ionic') { -%>
        map(loading => { this.loading = loading; loading.present() }),
        switchMap(() => login$
<% } -%>
        finalize(() => {
            this.loginForm.markAsPristine();
<% if (props.ui === 'ionic') { -%>
            this.loading.dismiss();
<% } -%>
        })
    );

captaincaius replied:
I tried it, and it worked great and looked great ( https://github.com/ngx-rocket/generator-ngx-rocket/blob/09e208569fc7fe47b22c73164e1afa053dc2ec47/generators/app/templates/src/app/login/__auth._login.component.ts )

...And all tests passed locally...

...And then e2e on CI started failing intermittently, and I couldn't reproduce it locally.

It looks like a bug in ionic to me that my quick and dirty promise code was somehow hiding, because there's no reason this version should be failing, because either way protractor is supposed to wait until the zone stabilizes, and both cases should stabilize only after BOTH the present() is done and the new component has fully loaded.

I couldn't investigate it further because I couldn't reproduce it, but something's gotta be up with the zone handling in either the loading controller or what they're doing with the router.

I'll try to find time to reproduce it with a simpler test case so I can submit it.

Anyway, I went back to the ugly version to confirm that this was indeed the variable.  I'll roll it back to the failing version later.

captaincaius replied:
FYI in case you friends are interested ;)

I was finally able to reproduce it locally with only promises and made a simple test case off of ionic's own official starters...  I'll be submitting this to ionic on Monday.

https://github.com/captaincaius/loadingtestcase/pull/1/files

captaincaius replied:
heh... looks like I pasted the PR diff, not the actual issue :P.

https://github.com/ionic-team/ionic/issues/15629

still no response yet on the issue.  I think I'll try to harass someone on ionic slack if I can't figure it out.  Debugging into minified transpiled ionic/core code is a 0/10 on the fun scale :P.

captaincaius replied:
comment added with link to commit to revert when loading controller works better
done :)

captaincaius commented:
Thanks for all the extra time you folks invested to review my first PR here.  I think I've addressed all your feedback now, so it's ready for re-review whenever you have time.

I was planning on getting it visually fully back to where it was on ionic3, but it seems there's still a lot of changes happening on ionic, so I decided it would be better to deal with that when it stabilizes a bit more.

I'll reply to all the comments individually so it's easier for you folks to confirm what's resolved and what's not :).

captaincaius commented on generators/app/templates/.editorconfig:
```diff
@@ -8,7 +8,7 @@ indent_size = 2
 trim_trailing_whitespace = true
 end_of_line = lf
-max_line_length = 120
+max_line_length = 140
```
reverted :)

captaincaius commented on generators/app/templates/.stylelintrc:
```diff
@@ -33,6 +39,6 @@
     "unit-whitelist": ["px", "%", "em", "rem", "vw", "vh", "deg", "s"],
     "max-empty-lines": 2,
-    "max-line-length": 120
+    "max-line-length": 140
```
reverted :)

captaincaius commented on generators/app/templates/_.gitignore:
```diff
@@ -48,6 +48,8 @@ testem.log
 /src/translations/template.*
 /src/environments/.env.*
+.ionic/
+.sourcemaps/
```
reverted :)

captaincaius commented on generators/app/templates/__cordova._config.xml:
```diff
@@ -117,6 +117,7 @@
 
 <% } -%>
   <!-- Common to all platforms -->
+  <preference name="ScrollEnabled" value="false"/>
```
reverted :)

captaincaius commented on generators/app/templates/__ionic._ionic.config.json:
```diff
@@ -0,0 +1,7 @@
+    "cordova": {}
+  },
+  "type": "angular"
+}
```
reverted :)

captaincaius commented on generators/app/templates/src/app/login/__auth+ionic.login.component.html:
```diff
@@ -2,50 +2,54 @@
-            </div>
-            <ion-list>
+            <ion-text color="danger" [hidden]="!error || loginForm.dirty">
+              <div padding translate>
```
done :)

captaincaius commented on generators/app/templates/src/app/login/__auth+ionic.login.component.html:
```diff
@@ -2,50 +2,54 @@
-                Username is required
-              </div>
+              <ion-text color="danger" [hidden]="loginForm.get('username').valid || loginForm.get('username').untouched">
+                <div padding translate>
```
done :)

captaincaius commented on generators/app/templates/src/app/login/__auth+ionic.login.component.html:
```diff
@@ -2,50 +2,54 @@
-              <ion-item padding-top no-lines>
-                <ion-label><span translate>Remember me</span></ion-label>
+              <ion-text color="danger" [hidden]="loginForm.get('password').valid || loginForm.get('password').untouched">
+                <div padding translate>
```
done :)

captaincaius commented on generators/app/templates/_angular.json:
```diff
@@ -114,6 +123,32 @@
+            }
+          }
+        }
+<% } -%>
```
reverted :)

captaincaius commented on generators/app/templates/_angular.json:
```diff
@@ -141,5 +176,18 @@
+      "styleext": "scss"
+    }
+  },
+<% } -%>
```
reverted :)

captaincaius commented on generators/app/templates/_package.json:
```diff
@@ -75,49 +70,56 @@
+    "@angular-devkit/core": "~0.7.5",
+    "@angular-devkit/schematics": "~0.7.5",
     "@biesbjerg/ngx-translate-extract": "^2.3.4",
+    "@ionic/ng-toolkit": "^1.0.0",
```
removed :)

captaincaius commented on generators/app/templates/_package.json:
```diff
@@ -75,49 +70,56 @@
     "@types/lodash": "^4.14.108",
-    "@types/node": "^8.10.13",
-    "codelyzer": "^4.3.0",
+    "@types/node": "~10.9.2",
```
fixed :)

captaincaius commented on generators/app/templates/e2e/src/_app.e2e-spec.ts:
```diff
@@ -6,20 +6,18 @@ import { AppPage } from './app.po';
 
 <% if (props.auth) { -%>
   it('should display login page and login into app', () => {
-    page.navigateTo();
```
ditto above - pending other PR

captaincaius commented on generators/app/templates/e2e/src/_app.e2e-spec.ts:
```diff
@@ -6,20 +6,18 @@ import { AppPage } from './app.po';
 
 <% } -%>
   it('should display hello message', () => {
-    page.navigateTo();
```
ditto above - pending other PR

captaincaius commented on generators/app/templates/src/_index.html:
```diff
@@ -6,6 +6,7 @@
   <base href="/"/>
 <% if (props.target.includes('cordova')) { -%>
   <meta name="viewport" content="viewport-fit=cover, width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no"/>
+  <meta name="format-detection" content="telephone=no"/>
```
reverted :)

captaincaius commented on generators/app/templates/src/_index.html:
```diff
@@ -15,6 +16,11 @@
+
+  <!-- add to homescreen for ios -->
+  <meta name="apple-mobile-web-app-capable" content="yes"/>
+  <meta name="apple-mobile-web-app-status-bar-style" content="black"/>
```
moved and changed to default :), although now I'm wondering if I should have added them to begin with...

captaincaius commented on generators/app/templates/src/app/_app.component.spec.ts:
```diff
@@ -2,12 +2,17 @@ import { TestBed, async } from '@angular/core/testing';
+import { RouteReuseStrategy } from '@angular/router';
+<%   if (props.target.includes('cordova')) { -%>
+import { Platform } from '@ionic/angular';
+<%   } -%>
```
done :)

captaincaius commented on generators/app/templates/src/app/_app.component.ts:
```diff
@@ -119,7 +109,6 @@ export class AppComponent implements OnInit {
 
   private onCordovaReady() {
     if (window['cordova']) {
-      this.keyboard.hideKeyboardAccessoryBar(true);
```
re-added using cordova-plugin-ionic-keyboard instead :)

captaincaius commented on generators/app/templates/src/app/_app.module.ts:
```diff
@@ -89,31 +89,14 @@ import { AppRoutingModule } from './app-routing.module';
-<%   } else { -%>
-    { provide: LocationStrategy, useClass: HashLocationStrategy }
-<%   } -%>
-<% } -%>
```
fixed :)

captaincaius commented on generators/app/templates/src/app/core/logger.service.ts:
```diff
@@ -83,6 +83,7 @@ export class Logger {
    * Works the same as console.log().
    */
   info(...objects: any[]) {
+    // tslint:disable-next-line:no-console
```
rule removed :)

captaincaius commented on generators/app/templates/src/app/login/__auth+ionic.login.component.html:
```diff
@@ -2,50 +2,54 @@
-              </div>
-              <ion-item padding-top no-lines>
-                <ion-label><span translate>Remember me</span></ion-label>
+              <ion-text color="danger" [hidden]="loginForm.get('password').valid || loginForm.get('password').untouched">
```
reverted :)

captaincaius commented on generators/app/templates/src/app/login/__auth._login.component.ts:
```diff
@@ -38,28 +38,37 @@ export class LoginComponent implements OnInit {
 <% if (props.ui === 'ionic') { -%>
-    const loading = this.loadingController.create();
-    loading.present();
+  async login() {
```
all async/await removed in favor of promises :)

captaincaius commented on generators/app/templates/src/app/settings/__ionic-tabs._settings.component.ts:
```diff
@@ -43,12 +44,12 @@ export class SettingsComponent implements OnInit {
+        header: this.translateService.instant('Change language'),
         inputs: this.i18nService.supportedLanguages.map(language => ({
-          type: 'radio',
+          type: 'radio' as TextFieldTypes,
```
I'd think so too, but they made TextFieldTypes a union of literal types ( https://github.com/ionic-team/ionic/blob/master/core/src/interface.d.ts#L34 )

Oddly enough, I'd think typescript would be okay with just type: 'radio', but for some reason I get this:

```
ERROR in src/app/settings/settings.component.ts(43,7): error TS2345: Argument of type '{ header: any; inputs: { type: string; name: string; label: string; value: string; checked: boole...' is not assignable to parameter of type 'AlertOptions'.
  Types of property 'inputs' are incompatible.
    Type '{ type: string; name: string; label: string; value: string; checked: boolean; }[]' is not assignable to type 'AlertInput[]'.
      Type '{ type: string; name: string; label: string; value: string; checked: boolean; }' is not assignable to type 'AlertInput'.
        Types of property 'type' are incompatible.
          Type 'string' is not assignable to type '"number" | "text" | "date" | "email" | "password" | "search" | "tel" | "url" | "checkbox" | "radio"'.
```

:D 'radio' looks a lot like 'radio' to me. lol.

sinedied replied:
just `radio` should definitely work, I guess they have to iron out some strange thing here 😄

captaincaius commented on generators/app/templates/src/app/settings/__ionic-tabs._settings.component.ts:
```diff
@@ -65,7 +66,7 @@ export class SettingsComponent implements OnInit {
-      })
-      .present();
+      });
+    await alertController.present();
```
de-async/await'ed ;)

sinedied commented:
Sorry for the delay, I just started my paternity leave I'll try to find some time to review your changes and merge the PR now 😄 

Thanks for your amazing work and reactivity!

sinedied commented on generators/app/templates/src/app/_app.component.ts:
```diff
@@ -29,6 +26,10 @@ import { Angulartics2GoogleAnalytics } from 'angulartics2/ga';
 
+<% if (props.target.includes('cordova')) { -%>
+declare var Keyboard: any;
+
```
no need to do that see ⬇️

captaincaius replied:
ah ok will do :).  I didn't know which option you'd dislike more :P. haha.

captaincaius replied:
done :)

sinedied commented on generators/app/templates/src/app/_app.component.ts:
```diff
@@ -119,7 +113,7 @@ export class AppComponent implements OnInit {
   private onCordovaReady() {
     if (window['cordova']) {
-      this.keyboard.hideKeyboardAccessoryBar(true);
+      Keyboard.hideFormAccessoryBar(true);
```
use `window['Keyboard']` to avoid the `declare var` on top.

Not pretty, but `@ionic-native/keyboard` should be updated soon anyways.

captaincaius replied:
donezo :)

sinedied commented on generators/app/templates/src/app/login/__auth._login.module.ts:
```diff
@@ -5,7 +5,8 @@ import { TranslateModule } from '@ngx-translate/core';
 import { NgbModule } from '@ng-bootstrap/ng-bootstrap';
 <% } else if (props.ui === 'ionic') { -%>
-import { IonicModule } from 'ionic-angular';
+import { FormsModule } from '@angular/forms';
```
Will fix in another PR them, I missed that one when migrating to form modules 😄

sinedied commented:
Looks good, almost there! 👍 
Once the the e2e PR is merged, the only remaining hard fix is the one mentioned by @luiza-cicone , waiting for Ionic's fix. In the meantime, I think we could merge this PR anyways, just add a comment to indicate it's waiting for a proper fix.

Good job overall! 😃

captaincaius commented:
Awesome! :) and thanks for the compliments!

So I think what's left for the next commit is:

- update from 6.x branch
- comment describing why we have the hard-to-follow async/loading stuff (and maybe a pointer to what to change it to once we can)
- kill the declare

LMK if I missed anything :)

Woohoo!  Getting close!

captaincaius commented:
I think this upcoming release will be a good time to update this :).  Although there's no RTL support yet, styling seems to have had lots of updates.

Also, although my issue I posted about the loader's ngZone got buried, I saw some recent commits on master about zone handling and also event handling.  It's possible it got fixed.

captaincaius commented:
I'll also update that 6.x release issue about remaining ionic4-related blockers this week while I'm in context too :) sorry for the silence.  Got sick twice and also had to do some traveling :(

captaincaius commented:
I believe I addressed everything here, so as soon as you (and CI) approve, I think this is ready to merge \o/

I'll update that other issue with any loose ends this revealed for 6.x :)

sinedied commented:
So far so good @captaincaius , it's time for merge 😃 
Thanks again for this huge work 👍
